### PR TITLE
feat(agent-core): record llm request trace in wire.jsonl

### PR DESCRIPTION
**Pull request by RealKai42**
## Related Issue

No linked issue — this implements an internally discussed design for session data trajectories. Problem statement below.

## Problem

`wire.jsonl` is the per-agent record log and the source of truth for resume/replay, but it did not capture everything sent to the model. It stored the rendered system prompt (`config.update`) and the full context messages, while three things were invisible: the tool definitions actually sent in the top-level `tools[]` (builtin and MCP schemas are rebuilt from live code on resume, so the historical bytes were unrecoverable across version upgrades), the effective per-request parameters (completion budget cap, env sampling overrides, thinking effort, beta API, tool-select disclosure state — several of which come from env/catalog and cannot be reconstructed after the fact), and the raw MCP `tools/list` results (held only in memory, lost on disconnect). Retry attempts, strict wire-compliance resends, and compaction's shrink-and-retry rounds left no trace at all, even though each is a real billed request.

For debugging and data-trajectory preservation we want every outbound request to be reconstructable from the wire log alone, at the logical-request level (system prompt + messages + tools + effective params). Byte-level HTTP body capture is explicitly out of scope and left to a future opt-in trace layer.

## What changed

Three observability record types, written by default:

- **`llm.tools_snapshot`** — content-addressed snapshot of the top-level tools table exactly as sent (after the `deferred` strip, mirroring `generate()`), written once per unique table. Under progressive tool disclosure the top level is byte-stable by design, so `select_tools` loads do not churn snapshots; loaded MCP schemas already reach the log via message-level `tools` declarations, so the two mechanisms compose with zero duplication.
- **`llm.request`** — one record per outbound request (every retry attempt, strict resend, and compaction round) carrying `kind`, `turnStep`/`attempt`, a `projection: 'strict'` marker, the effective params whitelist (provider, model, thinking effort, env temperature/top_p, completion budget cap, betaApi, tool-select state), hash links to the system prompt and tools snapshot, and `messageCount`/`droppedCount` so compaction's shrinking slices are reconstructable. Secrets are never recorded (no apiKey/baseURL).
- **`mcp.tools_discovered`** — the server's verbatim `tools/list` result plus how the agent gated it (allow-list, name collisions), deduplicated per server by content hash. The connection manager now retains the raw definitions alongside the converted tools.

Why this fits the repository: the records contract in `records/types.ts` required every persisted record to have resume semantics. This PR makes the two record classes explicit — state records rebuild state; observability records only restore their write-dedup cursors on replay (persisted ≠ replayed, with an explicit whitelist in `restoreAgentRecord`). Recording happens at the single `Agent.generate` choke point, so loop steps and compaction share one code path; the `LLMRequestLogFields` side channel gains the new fields, and `chatWithRetry` now preserves caller-set fields instead of rebuilding them. No wire protocol bump: adding record kinds is explicitly covered by the versioning policy in `records/migration/index.ts`, older readers skip unknown kinds, and the migration rewrite path preserves them.

Size cost is bounded: snapshots are ~50–150KB × the number of unique tables (typically 1–3 per session, identity-cached so the per-step cost is a pointer comparison), `llm.request` is ~350–400B per request — small against a message-dominated log.

Also included: vis renders the new record kinds (the exhaustive renderer registry enforces this), the context projector whitelists them as non-context records, and the sessions docs (en/zh) mention the request trace.

Edge cases covered by tests: snapshot dedup across turns and across resume, table changes producing a new snapshot, inlined system prompt for non-config callers, retry field preservation, MCP discovery dedup on reconnect and after resume, allow-list changes re-recording, and top-level snapshot stability across `select_tools` loads. Existing wire snapshots were regenerated to include the new records.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/MoonshotAI/kimi-code/blob/main/CONTRIBUTING.md) document.
- [x] I have linked a related issue, or explained the problem above.
- [x] I have added tests that prove my feature works.
- [x] Ran `gen-changesets` skill, or this PR needs no changeset. (`.changeset/wire-request-trace.md`: `agent-core` + `kimi-code` minor.)
- [x] Ran `gen-docs` skill, or this PR needs no doc update. (One-line update to `docs/{en,zh}/guides/sessions.md`.)

### CODE REVIEW

File: packages/agent-core/src/agent/turn/kosong-llm.ts [hunk 1]
```diff
@@ -122,7 +123,14 @@ export class KosongLLM implements LLM {
...
       onRequestSent: markRequestSent,
       onStreamEnd: markStreamEnd,
-      requestLogFields: params.requestLogFields,
+      requestLogFields: {
+        ...params.requestLogFields,
+        maxTokens: appliedCompletionBudgetCap({
+          provider: this.provider,
+          budget: this.completionBudgetConfig,
+          capability: this.capability,
+        }),
```
[the remark this answers is not in the document]
Comment (RealKai42) [reply packages/agent-core/src/agent/turn/kosong-llm.ts]: Good catch — and the clamp is actually broader than large-context turns: with the default budget (`hardCap` unset, cap = model window) the Kimi provider's remaining-context sizing changes the wire value on nearly every non-empty-context request, and the OpenAI chat-completions ceiling / Anthropic model-default resolution diverge even on the first step.

Fixed in 32b2e647, taking the exposure route instead of recomputing the clamp client-side: each provider's `withMaxCompletionTokens` already computes the final cap internally, so it now stores that value on the returned clone as a readonly `maxCompletionTokens` field, and the trace recorder reads it from the effective provider at the `Agent.generate` choke point. The recorded `llm.request.maxTokens` is now exactly the wire value for all five adapters, and the previous side-channel recomputation is removed.
